image2tensor/src/image2tensor.rs

```diff
@@ -1,83 +1,93 @@
     width: u32,
     height: u32,
     precision: TensorType,
-    out_buff: &mut [u8]
-    ) -> i32 {
-    let raw_file = Reader::open(path);
-    let raw_file = match raw_file {
-        Ok(file) => file,
-        Err(e) => {
-            println!("Failed to open the file: {:?}", e);
-            return -1;
-        },
-    };
+    order: ColorOrder,
+) -> Result<Vec<u8>, String> {
+    // Open the file and create the reader.
+    let raw_file = Reader::open(path)
+        .or_else(|_| Err(format!("Failed to open the file: {:?}", path)))
+        .unwrap();
 
-    let decoded = raw_file.decode(); //.unwrap();
-    let decoded = match decoded {
-        Ok(pixels) => pixels,
-        Err(e) => {
-            println!("Failed to decode the image: {:?}", e);
-            return -1;
-        },
-    };
+    // Create the DynamicImage by decoding the image.
+    let decoded = raw_file
+        .decode()
+        .or_else(|_| Err(format!("Failed to decode the file: {:?}", path)))
+        .unwrap();
 
+    // Resize the image to the specified W/H and get an array of u8 RGB values.
     let dyn_img: DynamicImage = decoded.resize_exact(width, height, image::imageops::Triangle);
-    // Switch from RGB to BGR
-    let bgr_img = dyn_img.to_bgr8();
-
-    // Get an array of the pixel values
-    let raw_u8_arr: &[u8] = &bgr_img.as_raw()[..];
-    let bytes_per_pixel: usize = get_bytes_per_pixel(precision);
-    let total_bytes: i32 = (raw_u8_arr.len() * bytes_per_pixel) as i32;
-
-    // Create an array to hold the f32 value of those pixels
-    for i in 0..raw_u8_arr.len()  {
-
-        // Read the number as a f32 and break it into u8 bytes
-        let u8_f32: f32 = raw_u8_arr[i] as f32;
-        let u8_bytes = u8_f32.to_ne_bytes();
-
-        for j in 0..bytes_per_pixel {
-            out_buff[(i * bytes_per_pixel) + j] = u8_bytes[j];
-        }
+    let mut img_bytes = dyn_img.into_bytes();
 
+    // Get an array of the pixel values and return it.
+    match order {
+        ColorOrder::RGB => Ok(save_bytes(&img_bytes, precision)),
+        ColorOrder::BGR => Ok(save_bytes(rgb_to_bgr(&mut img_bytes), precision)),
     }
-
-    return total_bytes;
 }
 
-pub fn calculate_buffer_size (
-    width: u32,
-    height: u32,
-    precision: TensorType) -> usize {
-
+pub fn calculate_buffer_size(width: u32, height: u32, precision: TensorType) -> usize {
     let bytes_per_pixel = get_bytes_per_pixel(precision);
     let pixels: u32 = width * height * 3;
-    return pixels as usize * bytes_per_pixel;
+    pixels as usize * bytes_per_pixel
 }
 
-fn get_bytes_per_pixel (precision: TensorType) -> usize {
+// Save the bytes into the specified TensorType format.
+fn save_bytes(arr: &[u8], tt: TensorType) -> Vec<u8> {
+    let mut out: Vec<u8> = vec![];
+
+    for i in 0..arr.len() {
+        // Split out the bytes based on the TensorType.
+        let bytes: Vec<u8>;
+        bytes = match tt {
+            TensorType::F16 => (arr[i] as f32).to_ne_bytes().to_vec(),
```

Comment:
I think this is incorrect with respect to the `calculate_buffer_size` and `get_bytes_per_pixel`: those places assume `f16` takes up 2 bytes but here we end up pushing 4.